meshcapade_addon/operators.py

```diff
@@ -0,0 +1,1660 @@
+        object_eval = obj.evaluated_get(depsgraph)
+        mesh_from_eval = object_eval.to_mesh()
+
+        # Get vertices in world coordinates
+        matrix_world = obj.matrix_world
+        vertices_world = [matrix_world @ vertex.co for vertex in mesh_from_eval.vertices]
+        z_min = (min(vertices_world, key=lambda item: item.z)).z
+        object_eval.to_mesh_clear() # Remove temporary mesh
+
+        # Adjust height of armature so that lowest vertex is on ground plane.
+        # Do not apply new armature location transform so that we are later able to show loaded poses at their desired height.
+        armature.location.z = armature.location.z - z_min
+
+        return {'FINISHED'}
+
+
+class OP_UpdateJointLocations(bpy.types.Operator):
+    bl_idname = "object.update_joint_locations"
+    bl_label = "Update Joint Locations"
+    bl_description = ("You only need to click this button if you change the shape keys from the object data tab (not using the plugin)")
+    bl_options = {'REGISTER', 'UNDO'}
+
+    j_regressor_female = {}
+    j_regressor_male = {}
+    j_regressor_neutral = {}
+
+    @classmethod
+    def poll(cls, context):
+        try:
+            # Enable button only if mesh is active object
+            return ((context.object.type == 'MESH') and (context.object.parent.type == 'ARMATURE'))
+        except Exception: 
+            return False
+
+    def load_regressor(self, gender, betas, SMPL_version):
+        if betas == 10:
+            suffix = ""
+        elif betas == 300:
+            suffix = "_300"
+        elif betas == 400:
+            suffix = "_400"
+
+        else:
+            print(f"ERROR: No betas-to-joints regressor for desired beta shapes [{betas}]")
+            return (None, None)
+
+        regressor_path = os.path.join(PATH, "data", f"{SMPL_version}_betas_to_joints_{gender}{suffix}.json")
+        with open(regressor_path) as f:
+            data = json.load(f)
+            return (np.asarray(data["betasJ_regr"]), np.asarray(data["template_J"]))
+
+    def execute(self, context):
+        obj = bpy.context.object
+        bpy.ops.object.mode_set(mode='OBJECT')
+
+        SMPL_version = bpy.context.object['SMPL_version']
+
+        # SMPLH is missing the joint regressor so we just want to exit
+        if (SMPL_version == 'SMPLH'):
+            return {'CANCELLED'}
+
+        gender = bpy.context.object['gender']
+        joint_names = get_joint_names(SMPL_version)
+        num_joints = len(joint_names)
+
+        # Get beta shapes
+        betas = []
+        for key_block in obj.data.shape_keys.key_blocks:
+            if key_block.name.startswith("Shape"):
+                betas.append(key_block.value)
+        num_betas = len(betas)
+        betas = np.array(betas)
+
+        # this seems like it's loading up every possibilty every time you click this button
+        # TODO: determine what you need and then only load what you need?
+        # Cache regressor files on first call
```

Comment:
How many files would we need to cache? 
I imagine at least 3 versions for male,female,neutral would be necessary. 
Then 2 versions for each for smplx and supr. Probably even more if we also load each beta version (10,300,400) 
Seems like this could easily occupy a lot of memory very quickly. 
Would some sort of model cache with a fixed capacity make sense?

Reply:
Maybe the best thing to do is to cache all of the files as soon as blender starts. 
![image](https://github.com/Meshcapade/SMPL_blender_addon/assets/123560441/5b5b8f9c-fe8f-43e3-9c36-ad4672c39fb8)

It was loading up every gender for every joint regressor.  There used to be 2 different joint regressors for SMPLX but I removed one of them.  I have now changed it so that it only does it for the gender you're working on.  That makes it faster.  But what's really terrible though is that we call UpdateJointLocations every frame whenever you slide around the height or weight sliders.  So it's loading from that file every frame.

I don't know anything about caching stuff and using it.  If you can provide insight here, that would make it run buttery smooth.

Reply:
I checked a bit the timings and on my machine and loading of the regressors takes around 19ms and updating the joints takes 10ms. I can look more into caching this stuff and possibly shave off the 19ms, but it should still be fine for this initial release. For now we can leave the todo comments and make another PR later with the caching.